### PR TITLE
chainerによる二層FFNNのMNIST分類器(F.Linear版)

### PR DESCRIPTION
**Pull request by matsumishoki**
chainerによる二層FFNNのMNIST分類器(F.Linear版)を実装する．

### CODE REVIEW

File: mnist_chainer_neural_network_v2.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+from sklearn.cross_validation import train_test_split
+import matplotlib.pyplot as plt
+import time
+import copy
+import chainer.functions as F
+from chainer import Variable, FunctionSet
+from chainer.optimizers import SGD, Adam
+
+
+def error_and_accuracy(model, x_data, t_data):
```
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]: error ではなく loss に変えてください。

File: mnist_chainer_neural_network_v2.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            x = Variable(x_batch)
+            t = Variable(t_batch)
+
+            # 順伝播
+            a_z = linear_1(x)
+            z = F.tanh(a_z)
+            a_y = linear_2(z)
+            
+            # 逆伝播            
+            loss = F.softmax_cross_entropy(a_y, t)
```
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]:
```
a_z = linear_1(x)
z = F.tanh(a_z)
a_y = linear_2(z)
loss = F.softmax_cross_entropy(a_y, t)
```

ここまとめて、 `loss_and_accuracy()` を使えばいいと思います。

（ちなみに、 `linear_1(x)` ではなく `model.linear_1(x)` を使うべきでしたね。）
（ちなみに、この行自体は逆伝播の最初のステップではなく、順伝播の最後のステップです。）

File: mnist_chainer_neural_network_v2.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            model_best = copy.deepcopy(model)
+            epoch_best = epoch
+            valid_error_best = valid_error
+            valid_accuracy_best = valid_accuracy
+            print "epoch_best:", epoch_best
+            print "valid_error_best:", valid_error_best
+            print "valid_accuracy_best:", valid_accuracy_best
+            print
+    # 学習済みのモデルをテストセットで誤差と正解率を求める
+    test_error, test_accuracy = error_and_accuracy(model,
```
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]: ここって、どの時点のモデルで評価してることになりますか？

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py]: 逆伝播が終わってからのモデルです．

Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]: まあ逆伝搬（更新後）なんですが、何回目の更新後の時点のモデルですか？

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py]: 1回目のエポックでのモデルを基準として交差エントロピー誤差の値を小さくなれば153行目から158行目で保存するようにできています．そのため，エポック毎に評価しています．そのため，過学習されたモデルで評価されている可能性もあります．

File: mnist_chainer_neural_network_v2.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    print "[valid] Error (best):", valid_error_best
+    print "Best epoch :", epoch_best
+    print "Finish epoch:", epoch
+    print "Batch size:", batch_size
+    print "Learning rate:", learning_rate
+    print "dim_hidden:", dim_hidden
+    print
+
+    # wの可視化
+    print "|w_1_best|:", np.linalg.norm(model.linear_1.W.data)
```
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py:199]: 重みのノルムなどは、毎エポック表示したほうがいいです。
なぜなら、学習の進行をモニタリングするための手段の一つだからです。
例えば、重みが異常な大きさになってないかどうかをある程度確認できることになります。

現状では、訓練・検証の損失・正解率をモニタリングしてることになりますが、
これだけの情報では、単に異常が起きているということしか検出できません。
しかし重みのノルムを表示するようにしておけば、得られる情報が増えるため、
一般的に、デバッグやチューニングが楽になります。

重みやバイアスのノルム以外にも、その勾配のノルムもモニタリングするようにしたほうがいいです。

逆に、この行のように、学習が終わった時点で表示しても別に変じゃないですが、あまりありがたみはないってことになります。

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py:199]: 毎エポックに重みのノルムの表示を増やしたいと思います．
これ以上表示するものを増やすと逆に見難くなりそうなのですが．．．
今，毎エポックに表示しているものが
epoch: 
time_elapsed:
[train] Error: 
[train] Accuracy:
[valid] Error:
[valid] Accuracy: 
とグラフです
現時点でも出力されるデータを全て見るのは少し難しいです．
何か逆に減らした方がいいものはありますか？

Comment (ronekko) [inline mnist_chainer_neural_network_v2.py:199]: 基本的には、減らしたほうがいいものはないですね。可能な限り、目に見えるようにすべきです。
現状はプロットを損失と正解率との2回表示（2個のfigureとして出力）してますが、これが縦幅をかなりとってるので、subplotを使って1個のfigureに2個のプロットを横に並べて表示したら縦幅がだいぶ節約できます。

File: mnist_chainer_neural_network_v2.py (hunk 3 above)
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]: まあそうですね。正確に言えば、この行における `model` は、最後のエポックにおける更新後のモデルなので、せっかく `model_best` を保存していることが無意味になってしまっています。ここでの評価は `model` ではなく `model_best` でしないとvalidの意味がなくなってしまいます。

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py]: 確かにそうですね．変更します．

File: mnist_chainer_neural_network_v2.py (hunk 4 above)
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py:199]: 例えば、以下のように書くことができます。 historyがlogになってたり変数名が若干違いますが、適宜読み替えてください。

```
            # plot learning curves
            plt.subplot(1, 2, 1)
            plt.title("Loss")
            plt.plot(loss_train_log)
            plt.plot(loss_valid_log)
            plt.plot(loss_test_log)
            plt.legend(["train", "valid", "test"], loc="best")
            plt.ylim([0.0, 0.5])
            plt.grid()

            plt.subplot(1, 2, 2)
            plt.title("Accuracy")
            plt.plot(accuracy_train_log)
            plt.plot(accuracy_valid_log)
            plt.plot(accuracy_test_log)
            plt.legend(["train", "valid", "test"], loc="best")
            plt.ylim([0.9, 1.0])
            plt.grid()
            plt.tight_layout()
            plt.show()
            plt.draw()
```

`plt.subplot(1, 2, 1)` というのは、サブプロットを1行×2列に配置するうちの1個目という意味になります。その後は普通にプロット用の処理を書けばよく、2個目の処理を書く前には今度は `plt.subplot(1, 2, 2)` を呼んだ後で普通に書けばOKです。

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py:199]: すみません．ありがとうございます．
Matplotlib galleryのページが開けずに色々どうすればいいか考えながら困っていました．

Comment (ronekko) [inline mnist_chainer_neural_network_v2.py:199]: http://matplotlib.org/gallery.html ですよね。こっちの環境では普通に開くことができているようです。

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py:199]: そうですね
そのページだけ検索しても開けないです．

Comment (ronekko) [inline mnist_chainer_neural_network_v2.py:199]: Googleから開くことができないという意味ですか？たまにありますね。
だったら上のリンクから直接開いて、お気に入りにいれといてください。というか既にお気に入りに入っているのでは？

File: mnist_chainer_neural_network_v2.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+        # 検証データセットの交差エントロピー誤差と正解率を表示する
+        valid_loss, valid_accuracy = loss_and_accuracy(model,
+                                                         x_valid, t_valid)
+        print "[valid] Loss:", valid_loss.data
+        print "[valid] Accuracy:", valid_accuracy
+        loss_valid_history.append(valid_loss.data)
+        valid_accuracy_history.append(valid_accuracy)
+        print "|w_1|:", np.linalg.norm(model.linear_1.W.data)
+        print "|w_2|:", np.linalg.norm(model.linear_2.W.data)
```
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py:149]: bのノルムも表示するようにしたほうがいいです。

File: mnist_chainer_neural_network_v2.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            
+            w_1_grad_norm = np.linalg.norm(model.linear_1.W.grad)
+            w_1_grad_norms.append(w_1_grad_norm)
+            w_2_grad_norm = np.linalg.norm(model.linear_2.W.grad)
+            w_2_grad_norms.append(w_2_grad_norm)
+        
+        w_1_grad_mean = np.mean(w_1_grad_norms, dtype=np.float32)
+        print "w_1_grad_mean:", w_1_grad_mean
+        w_2_grad_mean = np.mean(w_2_grad_norms, dtype=np.float32)
+        print "w_2_grad_mean:", w_2_grad_mean
```
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]: 表示する場所を，Wのノルムとかを表示してるとこにまとめたほうがいいのでは．

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py]: 処理部分と表示部分で離すということですか？
それとも，処理部分と表示部分をまとめて移動させて表示部分を固めるということですか？

Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]: むしろ，2行に分けているものをまとめて，

```
print "w_1_grad_mean:", np.mean(w_1_grad_norms, dtype=np.float32)
```

のようにしたらどうですか？

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py]: 確かに表示するものと変数名は同じである必要はないですね．
ありがとうございます．
変更します．

File: mnist_chainer_neural_network_v2.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+
+    loss_history = []
+    train_accuracy_history = []
+    loss_valid_history = []
+    valid_accuracy_history = []
+    w_1_grad_norms = []
+    w_2_grad_norms = []
+    b_1_grad_norms = []
+    b_2_grad_norms = []
```
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]: ここの4行は不要になりました。

Comment (matsumishoki) [inline mnist_chainer_neural_network_v2.py]: 確かにそうですね．

File: mnist_chainer_neural_network_v2.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        
+    w_1_best = 0
+    w_2_best = 0
```
Comment (ronekko) [inline mnist_chainer_neural_network_v2.py]: ここの2行も不要です。
